File: src/Film/film.cpp
    #include "film.hpp"
   
    //Constructeur par Default
    Film::Film(){
        id = -30;
    }

    //Constructeur d'init    
    Film::Film(string line){
        
        //Replace all Invalide CharType
        UtilityLib::replace_all(line, "&", "&amp;");
        UtilityLib::replace_all(line, "<", "&lt;");
        UtilityLib::replace_all(line, "\"", "&#39;");

        vector<string> tokenizedString = UtilityLib::getTokens(line, L"‣");
        int t = tokenizedString.size();

        try{ id = stoi(tokenizedString[0]); }
        catch(exception &err){ throw "FilmInit.Error: id not an Int"; }

        
        title = tokenizedString[1];
        originalTitle = tokenizedString[2];
        releaseDate = UtilityLib::stodate(tokenizedString[3]);
        status = tokenizedString[4];
        

        try{ voteAverage = stod(tokenizedString[5]); } 
        catch(exception &err){ throw "FilmInit.Error: voteAverage not a Double."; }
        
        try{ voteCount = stoi(tokenizedString[6]); }
        catch(exception &err){ throw "FilmInit.Error: voteCount not an Int.";}

        try{ runTime = stoi(tokenizedString[7]); }
        catch(exception &err){ 
            this->runTime = -1;
        }

        certification = tokenizedString[8];        
        posterPath = tokenizedString[9];

        try{ budget = stod(tokenizedString[10]); } 
        catch(exception &err){ throw "FilmInit.budget not a Double."; }

        tag = tokenizedString[11];

        if(tokenizedString[12].compare("") != 0){
            genres = Genre::stoGs(tokenizedString[12]);
        }
        
        if(tokenizedString[13].compare("") != 0){
            directors = Director::stoDs(tokenizedString[13]);
        }
        
        if(tokenizedString[14].compare("") != 0){
            if(t == 15){
                actors = Actor::stoAs(tokenizedString[14]);
            }
        }
    }

    //Constructeur de copievoid setIdGroupes(const Liste<int> &idG);
    Film::Film(const Film &e){
        *this = e;
    }

    //Deconstructeur
    Film::~Film(){

    }

    //Affiche un tuple de Film:
    void Film::Affiche(){
        cout << *this << endl;
    }

    //OPPERATOR SURCHARGE
    //=
    Film& Film::operator=(const Film& t2){
        setId(t2.getId());
        setTitle(t2.getTitle());
        setOriginalTitle(t2.getOriginalTitle());
        setRealeseDate(t2.getRealeseDate());
        setStatus(t2.getStatus());
        setVoteAverage(t2.getVoteAverage());
        setVoteCount(t2.getVoteCount());
        setRunTime(t2.getRunTime());
        setCertification(t2.getCertification());
        setPosterPath(t2.getPosterPath());
        setBudget(t2.getBudget());
        setTag(t2.getTag());
        setGenres(t2.getGenres());
        setDirectors(t2.getDirectors());
        setActors(t2.getActors());
        return *this;
    }


    //<<
    std::ostream& operator<<(std::ostream& s, const Film& e){
        if(e.getId() == -1){
            return s;
        }
        s << "\t<Film Id=\"" << e.getId() << "\" PosterPath=\"" << e.getPosterPath() << "\" >" << endl;
        
        s << "\t\t<Title OriginalTitle=\""<<e.getOriginalTitle()<<"\" >" << e.getTitle() << "</Title>"<< endl;

        if(e.getRealeseDate().tm_year != -1){
            s << "\t\t<Time Day=\"" << e.getRealeseDate().tm_mday <<"\" Month=\""<< e.getRealeseDate().tm_mon<<"\" Year=\""<< e.getRealeseDate().tm_year <<"\"/>"<<endl;
        }
        
        s << "\t\t<Status>" << e.getStatus() << "</Status>" << endl;
        
        if(e.getVoteCount() != 0 ){
            s << "\t\t<Vote Count=\"" << e.getVoteCount() << "\">" << e.getVoteAverage() << "</Vote>" << endl;
        }
        else{
            s << "\t\t<Vote Count=\"\"></Vote>" << endl;
        }
        
        if(e.getRunTime() > 0){
            s << "\t\t<RunTime>" << e.getRunTime() << "</RunTime>" << endl;
        }
        else{
            s << "\t\t<RunTime></RunTime>" << endl;
        }
        
        s << "\t\t<Certification>" << e.getCertification() << "</Certification>" << endl;
        
        if(e.getBudget()>0.1){
            s << "\t\t<Budget>" << e.getBudget() << "</Budget>" << endl;
        }
        else{
            s << "\t\t<Budget></Budget>" << endl;
        }

        s << "\t\t<Tag>" << e.getTag() << "</Tag>" << endl;

        s << "\t\t<Genres>"<<endl;
        for(int i = 0 ; i<e.getGenres().size() ; i++){
            s <<"\t\t\t"<< e.getGenres()[i] << endl;
        }
        s << "\t\t</Genres>"<<endl;

        s << "\t\t<Directors>"<<endl;
        for(int i = 0 ; i<e.getDirectors().size() ; i++){
            s <<"\t\t\t"<< e.getDirectors()[i] << endl;
        }
        s << "\t\t</Directors>"<<endl;

        s << "\t\t<Actors>"<<endl;
        for(int i = 0 ; i<e.getActors().size() ; i++){
            s <<"\t\t\t"<< e.getActors()[i] << endl;
        }
        s << "\t\t</Actors>"<<endl;

        s << "\t</Film>" << endl;
        return s;
    }



    
    int Film::getId() const
    {
        return id;
    }
    
    void Film::setId(const int idP)
    {
        id = idP;   
    }
    
    string Film::getTitle() const
    {
        return title;
    }
    
    void Film::setTitle(const string s)
    {
        title = s;
    }
    
    string Film::getOriginalTitle() const
    {
        return originalTitle;
    }
    
    void Film::setOriginalTitle(const string s)
    {
        originalTitle = s;
    }
    
    tm Film::getRealeseDate() const
    {
        return releaseDate;
    }
    
    void Film::setRealeseDate(const tm d)
    {
        releaseDate = d;
    }
    
    string Film::getStatus() const
    {
        return status;
    }
    
    void Film::setStatus(const string s)
    {
        status = s;
    }
    
    double Film::getVoteAverage() const
    {
        return voteAverage;
    }
    
    void Film::setVoteAverage(const double d)
    {
        voteAverage = d;
    }
    
    int Film::getVoteCount() const
    {
        return voteCount;
    }
    
    void Film::setVoteCount(const int v)
    {
        voteCount = v;
    }
    
    int Film::getRunTime() const
    {
        return runTime;
    }
    
    void Film::setRunTime(const int run)
    {
        runTime = run;
    }
    
    string Film::getCertification() const
    {
        return certification;
    }
    
    void Film::setCertification(const string c)
    {
        certification = c;
    }
    
    string Film::getPosterPath() const
    {
        return posterPath;
    }
    
    void Film::setPosterPath(const string ppath)
    {
        posterPath = ppath;
    }
    
    double Film::getBudget() const
    {
        return budget;
    }
    
    void Film::setBudget(const double b)
    {
        budget = b;
    }
    
    string Film::getTag() const
    {
        return tag;
    }
    
    void Film::setTag(const string s)
    {
        tag = s;
    }
    
    vector<Genre> Film::getGenres() const
    {
        return genres;
    }
    
    void Film::setGenres(const vector<Genre> v)
    {
        genres = v;
    }
    
    vector<Director> Film::getDirectors() const
    {
        return directors;
    }
    
    void Film::setDirectors(const vector<Director> d)
    {
        directors = d;
    }
    
    vector<Actor> Film::getActors() const
    {
        return actors;
    }
    
    void Film::setActors(const vector<Actor> a)
    {
        actors = a;
    }

File: SAX_Validator/src/Films/Film.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package Films;

import java.util.Collection;
import java.util.Iterator;
import java.util.Vector;

/**
 *
 * @author student
 */
public class Film implements Comparable<Film>, Cloneable {

    /**
     * @return the OriginalTitle
     */
    public String getOriginalTitle() {
        return OriginalTitle;
    }

    /**
     * @param OriginalTitle the OriginalTitle to set
     */
    public void setOriginalTitle(String OriginalTitle) {
        this.OriginalTitle = OriginalTitle;
    }

    /**
     * @return the Id
     */
    public int getId() {
        return Id;
    }

    /**
     * @param Id the Id to set
     */
    public void setId(int Id) {
        this.Id = Id;
    }

    /**
     * @return the Title
     */
    public String getTitle() {
        return Title;
    }

    /**
     * @param Title the Title to set
     */
    public void setTitle(String Title) {
        this.Title = Title;
    }

    /**
     * @return the Path
     */
    public String getPath() {
        return Path;
    }

    /**
     * @param Path the Path to set
     */
    public void setPath(String Path) {
        this.Path = Path;
    }

    /**
     * @return the Date
     */
    public String getDate() {
        return Date;
    }

    /**
     * @param Date the Date to set
     */
    public void setDate(String Date) {
        this.Date = Date;
    }

    /**
     * @return the Status
     */
    public String getStatus() {
        return Status;
    }

    /**
     * @param Status the Status to set
     */
    public void setStatus(String Status) {
        this.Status = Status;
    }

    /**
     * @return the VoteCount
     */
    public int getVoteCount() {
        return VoteCount;
    }

    /**
     * @param VoteCount the VoteCount to set
     */
    public void setVoteCount(int VoteCount) {
        this.VoteCount = VoteCount;
    }

    /**
     * @return the VoteAverage
     */
    public float getVoteAverage() {
        return VoteAverage;
    }

    /**
     * @param VoteAverage the VoteAverage to set
     */
    public void setVoteAverage(float VoteAverage) {
        this.VoteAverage = VoteAverage;
    }

    /**
     * @return the RunTime
     */
    public int getRunTime() {
        return RunTime;
    }

    /**
     * @param RunTime the RunTime to set
     */
    public void setRunTime(int RunTime) {
        this.RunTime = RunTime;
    }

    /**
     * @return the Certif
     */
    public String getCertif() {
        return Certif;
    }

    /**
     * @param Certif the Certif to set
     */
    public void setCertif(String Certif) {
        this.Certif = Certif;
    }

    /**
     * @return the Budget
     */
    public int getBudget() {
        return Budget;
    }

    /**
     * @param Budget the Budget to set
     */
    public void setBudget(int Budget) {
        this.Budget = Budget;
    }

    /**
     * @return the Summary
     */
    public String getSummary() {
        return Summary;
    }

    /**
     * @param Summary the Summary to set
     */
    public void setSummary(String Summary) {
        this.Summary = Summary;
    }

    /**
     * @return the Genres
     */
    public Vector<Genre> getGenres() {
        return Genres;
    }

    /**
     * @param Genres the Genres to set
     */
    public void setGenres(Vector<Genre> Genres) {
        this.Genres = Genres;
    }

    /**
     * @return the Directors
     */
    public Vector<Director> getDirectors() {
        return Directors;
    }

    /**
     * @param Directors the Directors to set
     */
    public void setDirectors(Vector<Director> Directors) {
        this.Directors = Directors;
    }

    /**
     * @return the Actors
     */
    public Vector<Actor> getActors() {
        return Actors;
    }

    /**
     * @param Actors the Actors to set
     */
    public void setActors(Vector<Actor> Actors) {
        this.Actors = Actors;
    }

    public Film(int Id, String path) {
        this.Id = Id;
        this.setPath(path);
    }

    public Film() {
        Id = -1;
    }
    
     @Override
    public Object clone() throws CloneNotSupportedException {
        Film copy = new Film();
        
        copy.setActors(Actors);
        copy.setBudget(Budget);
        copy.setCertif(Certif);
        copy.setDate(Date);
        copy.setDirectors(Directors);
        copy.setGenres(Genres);
        copy.setId(Id);
        copy.setOriginalTitle(OriginalTitle);
        copy.setPath(Path);
        copy.setRunTime(RunTime);
        copy.setStatus(Status);
        copy.setSummary(Summary);
        copy.setTitle(Title);
        copy.setVoteAverage(VoteAverage);
        copy.setVoteCount(VoteCount);
        
        return copy;
    }
   
    
    private int Id;
    private String Title;
    private String OriginalTitle;
    private String Path;
    private String Date;
    private String Status;
    private int VoteCount;
    private float VoteAverage;
    private int RunTime;
    private String Certif;
    private int Budget;
    private String Summary;
    private Vector<Genre> Genres;
    private Vector<Director> Directors;
    private Vector<Actor> Actors;

    @Override
    public String toString() {
        Iterator i = Genres.iterator();
        String msg;
        
        msg = "Film{" + "Id=" + Id + ", Title=" + Title + ", Path=" + Path + ", Date=" + Date + ", Status=" + Status + ", VoteCount=" + VoteCount + ", VoteAverage=" + VoteAverage + ", RunTime=" + RunTime + ", Certif=" + Certif + ", Budget=" + Budget + ", Summary=" + Summary + 
                ", Genres=(";
        while(i.hasNext()){
            Genre g = (Genre) i.next();
            msg+= " " + g.toString() + ", ";
        }
        
        msg +=") , Directors=(";
        
        i = Directors.iterator();
        while(i.hasNext()){
            Director g = (Director) i.next();
            msg+= " " + g.toString() + ", ";
        }
                
        msg +="), Actors=(";
        
        i = Actors.iterator();
        while(i.hasNext()){
            Actor g = (Actor) i.next();
            msg+= " " + g.toString() + ", ";
        }
        
        msg += ")}";
                
        return msg;        
    }

    @Override
    public int compareTo(Film o) {
        if(o == null) return +1;
        
        if(this.getVoteAverage() < o.getVoteAverage()){
            return +1;
        }
        if(this.getVoteAverage() > o.getVoteAverage()){
            return -1;
        }
        return 0;
    }
}







